internal/controllers/extension_controller_test.go

```diff
@@ -0,0 +1,210 @@
+					},
+				},
+			},
+		},
+	}
+
+	for _, e := range badExtensions {
+		ext := e.DeepCopy()
+		require.Error(t, cl.Create(ctx, ext), fmt.Sprintf("Failed on %q", e.ObjectMeta.GetName()))
+	}
+
+	invalidExtensions := []ocv1alpha1.Extension{
+		{
+			ObjectMeta: metav1.ObjectMeta{Name: "no-source", Namespace: extKey.Namespace},
+			Spec: ocv1alpha1.ExtensionSpec{
+				ServiceAccountName: "default",
+			},
+		}, {
+			ObjectMeta: metav1.ObjectMeta{Name: "no-package", Namespace: extKey.Namespace},
+			Spec: ocv1alpha1.ExtensionSpec{
+				Source:             ocv1alpha1.ExtensionSource{},
+				ServiceAccountName: "default",
+			},
+		},
+	}
+
+	for _, e := range invalidExtensions {
+		ext := e.DeepCopy()
+		require.NoError(t, cl.Create(ctx, ext), fmt.Sprintf("Create failed on %q", e.GetObjectMeta().GetName()))
+		name := types.NamespacedName{Name: e.GetObjectMeta().GetName(), Namespace: e.GetObjectMeta().GetNamespace()}
+
+		cl, reconciler := newClientAndExtensionReconciler(t)
+		res, err := reconciler.Reconcile(ctx, ctrl.Request{NamespacedName: name})
+		require.Equal(t, ctrl.Result{}, res)
+		require.NoError(t, err)
+
+		require.NoError(t, cl.Get(ctx, name, ext), fmt.Sprintf("Get failed on %q", e.ObjectMeta.GetName()))
+		cond := apimeta.FindStatusCondition(ext.Status.Conditions, ocv1alpha1.TypeResolved)
+		require.NotNil(t, cond, fmt.Sprintf("Get condition failed on %q", ext.ObjectMeta.GetName()))
+		require.Equal(t, metav1.ConditionUnknown, cond.Status, fmt.Sprintf("Get status check failed on %q", ext.ObjectMeta.GetName()))
+		require.Equal(t, ocv1alpha1.ReasonResolutionUnknown, cond.Reason, fmt.Sprintf("Get status reason failed on %q", ext.ObjectMeta.GetName()))
+		require.Equal(t, "validation has not been attempted as spec is invalid", cond.Message)
+	}
+
+	require.NoError(t, cl.DeleteAllOf(ctx, &ocv1alpha1.Extension{}, client.InNamespace(extKey.Namespace)))
+	require.NoError(t, cl.Delete(ctx, namespace))
+}
+
+func TestExtensionInvalidSemverPastRegex(t *testing.T) {
+	cl, reconciler := newClientAndExtensionReconciler(t)
+	ctx := context.Background()
+	t.Log("When an invalid semver is provided that bypasses the regex validation")
+	extKey := types.NamespacedName{
+		Name:      fmt.Sprintf("extension-test-%s", rand.String(8)),
+		Namespace: fmt.Sprintf("namespace-%s", rand.String(8)),
+	}
+	namespace := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: extKey.Namespace,
+		},
+	}
+	require.NoError(t, cl.Create(ctx, namespace))
+
+	t.Log("By injecting creating a client with the bad clusterextension CR")
+	pkgName := fmt.Sprintf("exists-%s", rand.String(6))
+	extension := &ocv1alpha1.Extension{
+		ObjectMeta: metav1.ObjectMeta{Name: extKey.Name, Namespace: extKey.Namespace},
+		Spec: ocv1alpha1.ExtensionSpec{
+			Source: ocv1alpha1.ExtensionSource{
+				Package: &ocv1alpha1.ExtensionSourcePackage{
+					Name:    pkgName,
+					Version: "1.2.3-123abc_def", // bad semver that matches the regex on the CR validation
```

Comment:
This passes our CRD schema validation, but does not parse with the masterminds parser? If that's the case, how hard would it be to update the CRD pattern to disallow this?

Reply:
This was from cluster extension (originally operator), we might have to look at the original PR to understand why... my recollection was complexity of the regex, maybe?

Reply:
Looking further into the test, it purposefully bypasses CR validation, and exercises the validation code.